test/lib/jdk/test/lib/apps/LingeredApp.java

```diff
@@ -263,23 +265,13 @@ public void waitAppTerminate() {
      * @param timeout timeout in seconds
      * @throws java.io.IOException
      */
-    public void waitAppReady(long timeout) throws IOException {
+    public void waitAppReadyOrCrashed(long timeout) throws IOException {
         // adjust timeout for timeout_factor and convert to ms
         timeout = Utils.adjustTimeout(timeout) * 1000;
         long here = epoch();
         while (true) {
-            long epoch = epoch();
-            if (epoch - here > timeout) {
-                throw new IOException("App waiting timeout");
-            }
-
-            // Live process should touch lock file every second
-            long lm = lastModified(lockFileName);
-            if (lm > lockCreationTime) {
-                break;
-            }
-
-            // Make sure process didn't already exit
+            // Make sure process didn't already exit:
+            // check this now, and immediately after sleeping for spinDelay each loop.
```

Comment:
I don't see how this reordering of the code makes any difference.

Reply:
My thinking was that we check for exceeding timeout at the start of the loop, so if we do loop, we have just slept for spinDelay, and not checked the conditions for successful exit of the loop.  So we lose 1 second of timeout.
Now I see my error is that I should move the timeTaken check further down, after we check the lock file, so we do all success checking, then timeout checking, then sleep for spinDelay.